### DISCUSSION TITLE
BatchSampler mistakes batch size as channel number

### DISCUSSION THREAD

**Discussion by lxmedai** [Project-MONAI/MONAI#6613] [Q&A] [answered]
```
import torch
from torch.utils.data import SequentialSampler, BatchSampler, DataLoader

def data_loader(dataset, indices_subset, batch_size=5, drop_last=False):
    pin_memory = torch.cuda.is_available()
    return DataLoader(dataset, sampler=BatchSampler(SequentialSampler(indices_subset), 
                                  batch_size=batch_size, drop_last=drop_last), pin_memory=pin_memory)
```

I use the batch sampler for my data loader and it yields each batched data with size of (1, 5, 256, 256, 256) if I run the enumerator and check the data size like:
`for i, (X, _) in enumerate(dataloader):
         print(X.shape)  `
I however would like it to be size to be (5, 1, 256, 256, 256) as channel number should be 1 and the batch size should be 5 instead.
I can permute the data after calling emumerate to get right size but I don’t want to modify the standardized module of SupervisedTrainer (imported from monai.engines) for the downstream processing which I intend to use for my project. How should I customize the dataloader definition so that it will provide batched input data of the right size to a 3D classification model by itself and I don’t need to modify any imported packages which I use in the downstream? I thank anyone who would like to help out in advance!

**Comment by wyli** [comment]
the code example looks like a pure pytorch implementation, could you provide more details, I guess the `dataset` is some dataset + transforms from monai?

**Reply by lxmedai** [reply]
import h5py
from torch.utils.data import Dataset

class HDF5Dataset(Dataset):
    """Dataset to load data from the dataset .h5 files
    :param project_config: project configuration contains all necessary project-level information
    :type project_config: dict
    :param transform: transformation to apply to the images
    :type transform: list
    :param target_transform: transforms to add to each target/label
    :type target_transform: list
    :param max_size: maximum size of data to draw from
    :type max_size: int or None
    """

    def __init__(self, project_config, transform=[], target_transform=[], max_size=None):
        dir_path = os.path.join(project_config['OutputPath'], 'Dataset')
        file_name = project_config['DatasetName']
        super(HDF5Dataset, self).__init__()
        self.dir_path = dir_path
        self.inputs = self._load_h5_file_with_data(file_name, 'data')
        self.data_shape = tuple(self.inputs.attrs['DataShape'])
        self.labels = self._load_h5_file_with_data(file_name, 'labels')
        self.transform = Compose([self._from_numpy, self._reshape] + transform)
        self.target_transform = Compose([self._from_numpy, self._to_torch_int] + target_transform)
        self.max_size = max_size

    def __getitem__(self, index):
        inputs = self.transform(self.inputs[index])
        labels = self.target_transform(self.labels[index])
        return inputs, labels

    def __len__(self):
        return self.max_size if self.max_size else self.inputs.shape[0]

    def _load_h5_file_with_data(self, file_name, data_name):
        path = os.path.join(self.dir_path, file_name)
        file = h5py.File(path)
        data = file[data_name]
        return data

    def _reshape(self, tensor):
        # Reshape the flattened tensor in .h5 dataset file according to the specified data shape
        return torch.reshape(tensor, tuple([tensor.shape[0]]+list(self.data_shape)))

    def _from_numpy(self, tensor):
        return torch.from_numpy(tensor).float()

    def _to_torch_int(self, tensor):
        # transformation for labels
        return torch.reshape(tensor, (-1,)).type(torch.ByteTensor)


Here above is how the dataset is implemented (pure pytorch code as well). The stored dataset is in .h5 format and each image is stored as a flattened array (1*16777216) which will be reshaped to the original 3D shape (i.e. (256, 256, 256) ) using the stored 'DataShape' attribute associated with the image data.

**Reply by wyli** [reply]
I see, maybe in the original example you can try
```
DataLoader(..., batch_sampler=BatchSampler(SequentialSampler(...))
```
instead of **sampler**=BatchSampler, using the latter is treating a 5x256x256x256 array as one training item.

this is a generic pytorch question, I think if you ask in a pytorch forum you'll get more help.

**Reply by lxmedai** [reply]
I have posted the same question in the pytorch forum and I got the suggestion to set the batch_size of the DataLoader to None (not the BatchSampler) which I have tested before asking for advices, the data shape becomes naturally (5, 256, 256, 256), but then it creates a problem for my downstream using SupervisedTrainer (imported from monai.engines) and I get a RuntimeError complaining about the incompatibility with the weight size that input with 1 channel is expected but it gets 5 channels, which again is due to the error of mistaking batch size as channel number. Somehow the ignite engine recreates each batch to have the size of (1, 5, 256, 256, 256) through the internal run generator.

Assigning the BatchSampler to batch_sampler instead of sampler in the DataLoader constructor generates a ValueError saying "batch_sampler option is mutually exclusive with batch_size, shuffle, sampler, and drop_last".

**Reply by wyli** [reply]
> the data shape becomes naturally (5, 256, 256, 256), but then it creates a problem for my downstream using SupervisedTrainer (imported from monai.engines) and I get a RuntimeError complaining about the incompatibility with the weight size

I think that's the direction to go, if the `__getitem__` of the dataset instead returning a 1x256x256x256 array, then I think SupervisedTrainer and the network will receive shape(5, 1, 256, 256, 256), which should be compatible with 3D networks. Batchsize=5,numchannels=1,spatial dims=256,256,256.

**Reply by lxmedai** [reply]
How would you suggest to modify the code for the` __getitem__` method? Now the returned 'inputs' of the `__getitem__` method provides me with a batch instead of one single image.

**Reply by wyli** [reply]
I think you can start from this example and modify it accordingly:
```py
import numpy as np
from torch.utils.data import BatchSampler, DataLoader, Dataset, SequentialSampler

class MyDataset(Dataset):
    def __getitem__(self, index):
        return np.zeros((1, 256, 256, 256)), np.zeros((2, 256, 256, 256))

def data_loader(dataset, indices_subset, batch_size=5, drop_last=False):
    return DataLoader(
        dataset,
        batch_sampler=BatchSampler(SequentialSampler(indices_subset), batch_size=batch_size, drop_last=drop_last),
    )

for i, (x, y) in enumerate(data_loader(MyDataset(), range(50))):
    print(x.shape, y.shape)
```
```
torch.Size([5, 1, 256, 256, 256]) torch.Size([5, 2, 256, 256, 256])
torch.Size([5, 1, 256, 256, 256]) torch.Size([5, 2, 256, 256, 256])
```

it's a generic pytorch question, I'd suggest please post further questions to the pytorch forum, so that potentially more users may benefit from the Q&A.

**Reply by lxmedai** [reply, marked as answer]
I tested adding `inputs = torch.unsqueeze(inputs, 1)` after the line of `inputs = self.transform(self.inputs[index])`, and now it is working. I have previously also posted my question there and the current solution has also been posted in pytorch forum. Thanks for your help!
